packages/mdc-menu-surface/index.ts

```diff
@@ -0,0 +1,207 @@
+    this.foundation_.setAbsolutePosition(x, y);
+    this.setIsHoisted(true);
+  }
+
+  /** @param corner Default anchor corner alignment of top-left surface corner. */
+  setAnchorCorner(corner: Corner) {
+    this.foundation_.setAnchorCorner(corner);
+  }
+
+  setAnchorMargin(margin: Partial<MenuDistance>) {
+    this.foundation_.setAnchorMargin(margin);
+  }
+
+  getDefaultFoundation(): MDCMenuSurfaceFoundation {
+    // tslint:disable:object-literal-sort-keys
+    return new MDCMenuSurfaceFoundation({
+      addClass: (className) => this.root_.classList.add(className),
+      removeClass: (className) => this.root_.classList.remove(className),
+      hasClass: (className) => this.root_.classList.contains(className),
+      hasAnchor: () => !!this.anchorElement,
+      notifyClose: () => this.emit(MDCMenuSurfaceFoundation.strings.CLOSED_EVENT, {}),
+      notifyOpen: () => this.emit(MDCMenuSurfaceFoundation.strings.OPENED_EVENT, {}),
+      isElementInContainer: (el) => this.root_.contains(el),
+      isRtl: () => getComputedStyle(this.root_).getPropertyValue('direction') === 'rtl',
+      setTransformOrigin: (origin) => {
+        const propertyName = `${util.getTransformPropertyName(window)}-origin`;
+        this.root_.style.setProperty(propertyName, origin);
+      },
+
+      isFocused: () => document.activeElement === this.root_,
+      saveFocus: () => {
+        this.previousFocus_ = document.activeElement as HTMLElement | SVGElement;
+      },
+      restoreFocus: () => {
+        if (this.root_.contains(document.activeElement)) {
+          if (this.previousFocus_ && this.previousFocus_.focus) {
+            this.previousFocus_.focus();
+          }
+        }
+      },
+      isFirstElementFocused: () =>
+          this.firstFocusableElement_ ? this.firstFocusableElement_ === document.activeElement : false,
+      isLastElementFocused: () =>
+          this.lastFocusableElement_ ? this.lastFocusableElement_ === document.activeElement : false,
+      focusFirstElement: () =>
+          this.firstFocusableElement_ && this.firstFocusableElement_.focus && this.firstFocusableElement_.focus(),
+      focusLastElement: () =>
+          this.lastFocusableElement_ && this.lastFocusableElement_.focus && this.lastFocusableElement_.focus(),
+
+      getInnerDimensions: () => {
+        return {width: this.root_.offsetWidth, height: this.root_.offsetHeight};
+      },
+      getAnchorDimensions: () => this.anchorElement ? this.anchorElement.getBoundingClientRect() : null,
+      getWindowDimensions: () => {
+        return {width: window.innerWidth, height: window.innerHeight};
+      },
+      getBodyDimensions: () => {
+        return {width: document.body.clientWidth, height: document.body.clientHeight};
+      },
+      getWindowScroll: () => {
+        return {x: window.pageXOffset, y: window.pageYOffset};
+      },
+      setPosition: (position) => {
+        this.root_.style.left = 'left' in position ? `${position.left}px` : '';
```

Comment:
FYI, you might run into trouble here if closure decides to minify these properties, I would recommend this pattern instead:

```ts
this.root_.style.left = position.left == null ? `{position.left}px` : '';
```

(using `==` intentionally as lets `0` through but changes `null` and `undefined` to `''`)

Reply:
IIRC, Closure Compiler munges the property names of objects (unless you explicitly prevent it from doing so with an annotation or w/e), but it _doesn't_ rename `string` values like `'left'`.

So the object will get minified/obfuscated to something like `{a, b, c, d}`, whereas this line of code would still be testing for the existence of a property named `'left'` that no longer exists.

Using `.dotNotation` avoids this problem by allowing Closure to rename the property everywhere.

Reply:
Oh I see, I was thinking about this backwards then (thinking about avoiding it being obfuscated whatsoever rather than being obfuscated consistently).

I'm still a little unnerved about `!=` and am wondering if we can be consistent on one of `null` or `undefined` specifically?

Reply:
If this is indeed an issue, it is present in several places:

https://github.com/material-components/material-components-web/blob/86cfdadd4fc999f696d0b057cb80afa0e3061120/packages/mdc-menu-surface/foundation.ts#L260-L261

https://github.com/material-components/material-components-web/blob/86cfdadd4fc999f696d0b057cb80afa0e3061120/packages/mdc-menu-surface/foundation.ts#L266-L269

https://github.com/material-components/material-components-web/blob/86cfdadd4fc999f696d0b057cb80afa0e3061120/packages/mdc-menu-surface/foundation.ts#L449-L450

https://github.com/material-components/material-components-web/blob/86cfdadd4fc999f696d0b057cb80afa0e3061120/packages/mdc-menu-surface/foundation.ts#L459-L463

Reply:
It's also worth noting that this code already existed in this fashion both externally and internally prior to this conversion, so it would seem it hasn't caused an issue.

Reply:
After talking to Ken offline, we're not too worried about this because `MDCMenuFoundation` is already running with prod jscompiler minification/obfuscation settings.

Thanks for bringing this up, though!